dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/SubProcessTaskExecThread.java

```diff
@@ -112,6 +112,31 @@ private void updateParentProcessState(){
         this.processInstance.setState(parentProcessInstance.getState());
     }
 
+    /**
+     * wait for the establishment of sub-process-instance
+     * @throws InterruptedException
+     */
+    private void waitSubProcessEstablishment() throws InterruptedException {
+        if (subProcessInstance == null) {
+            while (Stopper.isRunning()) {
+                Thread.sleep(Constants.SLEEP_TIME_MILLIS);
+                if (setTaskInstanceState()) {
+                    break;
+                }
+            }
+        }
+    }
+
+    /**
+     * check if both task and sub-process are both finished
+     * (when recover a stopping instance, subProcessInstance may still in running by fault-tolerant)
+     */
+    private boolean isTaskInstanceFinished() {
+        boolean isSubProcessFinished = (subProcessInstance == null || subProcessInstance.getState().typeIsFinished());
```

Comment:
we have already checked the sub process is not null.

Reply:
It is used to ensure that `subProcessInstance.getState()` is valid, it may still be null at this position.

Reply:
After submitting the task by super.submit(), the subprocess will be submitted to the command list. At that time, the subprocess has not been established.